src/zenml/integrations/mlflow/steps/mlflow_deployer.py

```diff
@@ -111,98 +111,83 @@ def mlflow_model_deployer_step(
+            ),
+        )
         logger.info(
-            f"An MLflow model with name `{model_name}` was not "
-            f"trained in the current pipeline run. Reusing the existing "
-            f"MLflow model server."
+            f"MLflow deployment service started and reachable at:\n"
+            f"    {new_service.prediction_url}\n"
         )
-        if not service.is_running:
-            service.start(timeout)
-
-        # return the existing service
-        return service
-
-    # even when the deploy decision is negative, if an existing model server
-    # is not running for this pipeline/step, we still have to serve the
-    # current model, to ensure that a model server is available at all times
-    if not deploy_decision and existing_services:
+        logger.info("Stopping existing services...")
+        for existing_service in existing_services:
+            existing_service.stop(timeout)
+        return new_service
+
+    # Check whether an existing service can be reused
+    elif existing_services:
+        if not model_uri:
+            logger.info(
+                f"Skipping model deployment because an MLflow model with name "
+                f" `{model_name}` was not trained in the current pipeline run."
+            )
+        elif not deploy_decision:
+            logger.info(
+                "Skipping model deployment because the deployment decision "
+                "was negative."
+            )
         logger.info(
-            f"Skipping model deployment because the model quality does not "
-            f"meet the criteria. Reusing last model server deployed by step "
-            f"'{step_name}' and pipeline '{pipeline_name}' for model "
-            f"'{model_name}'..."
+            f"Reusing last model server deployed by step '{step_name}' of "
+            f"pipeline '{pipeline_name}' for model '{model_name}'..."
         )
-        # even when the deploy decision is negative, we still need to start
-        # the previous model server if it is no longer running, to ensure
-        # that a model server is available at all times
+        service = cast(MLFlowDeploymentService, existing_services[0])
         if not service.is_running:
             service.start(timeout)
         return service
 
-    # create a new model deployment and replace an old one if it exists
-    new_service = cast(
-        MLFlowDeploymentService,
-        model_deployer.deploy_model(
-            replace=True,
-            config=predictor_cfg,
-            timeout=timeout,
-        ),
-    )
-
-    logger.info(
-        f"MLflow deployment service started and reachable at:\n"
-        f"    {new_service.prediction_url}\n"
-    )
-
-    return new_service
+    # Log a warning if no service was deployed or found
+    elif deploy_decision and not model_uri:
+        logger.warning(
+            f"An MLflow model with name `{model_name}` was not "
+            f"logged in the current pipeline run and no running MLflow "
```

Comment:
Not sure that I get it: `model_uri` depends on `mlflow_run_id`, which we got from the MlFlow client based on the experiment or pipeline name.
First: I doubt that naming Experiment as Pipelines automatically is not well protected from collisions, but this is not related to this PR.
Second: if the user fed in the experiment name it will be confusing how Pipeline Run and the model not found in my experiment relate to one another.
So I suggest rephrasing this or making more flexible depending on user inputs.

Reply:
Yeah currently the MLflow deployer always assumes that you had an MLflow experiment tracker in a previous stack that logged the model as an MLflow artifact with name "model". We already know this UX is not ideal, refactoring this is actually the main point of the "Complete the model story" PRD is about AFAIK.

Reply:
Not really, I checked the code there and how I read this:
- User provides experiment name in params - search inside of it
- User skipped experiment name - search in experiment named pipeline name

To me, this logging is fine for use-case 2, but doesn't make sense for use-case 1.

Reply:
I didn't fully get that, but anyways I think we should fundamental discussions for the big refactoring coming soon; the goal of this PR was just to stop existing servers when deploying new models